packages/core/src/command/hl7v2-subscriptions/hl7v2-to-fhir-converter.ts

```diff
@@ -0,0 +1,324 @@
+  const { log } = out(`Hl7 to FHIR Lambda - cx: ${cxId}, pt: ${patientId}`);
+  log(`Running message from ${messageReceivedTimestamp}`);
+  const hl7Message = Hl7Message.parse(message);
+  const s3Client = getS3UtilsInstance();
+  const internalHl7RouteUrl = `${apiUrl}/${INTERNAL_HL7_ENDPOINT}`;
+  const internalGetPatientUrl = `${apiUrl}/${INTERNAL_GET_PATIENT_ENDPOINT}/${patientId}?cxId=${cxId}`;
+
+  const msgType = getHl7MessageTypeOrFail(hl7Message);
+  const shouldNotifyApi =
+    createMessageTypes.includes(msgType.triggerEvent) ||
+    updateMessageTypes.includes(msgType.triggerEvent);
+
+  const convertedBundle = convertHl7v2MessageToFhir({
+    hl7Message,
+    cxId,
+    patientId,
+    messageId,
+    timestampString: messageReceivedTimestamp,
+  });
+
+  const newBundle = await saturateConvertedBundle({
+    bundle: convertedBundle,
+    patientId,
+    cxId,
+    messageId,
+    messageReceivedTimestamp,
+    internalGetPatientUrl,
+    log,
+  });
+
+  const sharedUploadParams: Omit<StoreInS3Params, "fileName" | "payload"> = {
+    s3Utils: s3Client,
+    bucketName,
+    contentType: JSON_APP_MIME_TYPE,
+    log,
+    errorConfig: {
+      errorMessage: "Error uploading HL7 FHIR bundle to S3",
+      context: "convertHl7MessageToFhirAndUpload",
+      captureParams: {
+        patientId,
+        cxId,
+        messageReceivedTimestamp,
+      },
+      shouldCapture: true,
+    },
+  };
+
+  const newBundleFileName = buildHl7MessageFhirBundleFileKey({
+    cxId,
+    patientId,
+    timestamp: messageReceivedTimestamp,
+    messageId,
+    messageType: msgType.messageType,
+    messageCode: msgType.triggerEvent,
+  });
+
+  const newBundleUploadPromise = storeInS3WithRetries({
+    ...sharedUploadParams,
+    payload: JSON.stringify(newBundle),
+    fileName: newBundleFileName,
+  });
+
+  const combinedBundleFileName = buildHl7MessageCombinedBundleFileKey(cxId, patientId);
+  let combinedBundle = newBundle;
+  try {
+    const existingCombinedBundleRaw = await s3Client.downloadFile({
+      bucket: bucketName,
+      key: combinedBundleFileName,
+    });
+    const existingBundle = JSON.parse(existingCombinedBundleRaw.toString("utf-8"));
+    const mergedBundle = await mergeIncomingBundleIntoCombined({
+      cxId,
+      patientId,
+      existingBundle,
+      incomingBundle: newBundle,
+      log,
+    });
+    combinedBundle = mergedBundle;
+  } catch (err) {
+    log(`No existing combined bundle found. Creating a new one.`);
+    // intentionally not throwing
+  }
+
+  const combinedBundleUploadPromise = storeInS3WithRetries({
+    ...sharedUploadParams,
+    payload: JSON.stringify(combinedBundle),
+    fileName: combinedBundleFileName,
+    ...(sharedUploadParams.errorConfig
+      ? {
+          errorConfig: {
+            ...sharedUploadParams.errorConfig,
+            errorMessage: "Error uploading updated HL7 Combined FHIR bundle to S3",
+          },
+        }
+      : undefined),
+  });
```

Comment:
Adding this new resource to our existing FHIR bundle is a substantial new responsibility that the lambda is taking on, and could impact deliverability of messages. My understanding is bundle can be big, dedupe can take a while, right? if so, we really want to do this in a separate lambda. This is something to check w raf + thomas about. if the bundle DL and dedup takes 200ms its totally fine. Maybe even 5s or so and its fine. But if its taking 20s for like a P90 bundle, this is going to substantially affect message deliverability since at that point id guess a P99 bundle is gonna be a many minute long dedupe.

Reply:
I would say at a minimum, lets do this after attempting to deliver the webhook notification. But depending on how long this takes, we may need to invoke another lambda execution or have the initial S3 write trigger a lambda execution.